File: hash table/python/Hashtable.py
# hash table using python


class Hash:
    def __init__(self):
        self.hash_no = 0

    #  get a hash value for a given string
    def hash(self, input_string, max_size):
        for letter in range(len(input_string)):
            self.hash_no = self.hash_no + ord(input_string[letter])

        return self.hash_no % max_size


class HashTable:
    def __init__(self, size):
        self.hash_table = [None] * size
        self.size = size
        self.inserted = False
        self.hash = None

    # add a item to the hash table
    def add(self, key, value):
        self.hash = Hash()
        index = self.hash.hash(key, self.size)
        self.inserted = False

        if self.hash_table[index] is None:
            self.hash_table[index] = [[key, value]]
            self.inserted = True
        else:
            for arr in range(len(self.hash_table[index])):
                if self.hash_table[index][arr][0] == key:
                    self.hash_table[index][arr][1] = value
                    self.inserted = True

        if not self.inserted:
            self.hash_table[1].append([key, value])

    # remove a item to the hash table
    def remove(self, key):
        self.hash = Hash()
        index = self.hash.hash(key, self.size)

        if len(self.hash_table[index]) == 1 and self.hash_table[index][0][0] == key:
            del self.hash_table[index]
        else:
            for arr in range(len(self.hash_table[index])):
                if self.hash_table[index][arr][0] == key:
                    del self.hash_table[index][arr]
                    break

    # get a item from the hash table using the key
    def get(self, key):
        self.hash = Hash()
        index = self.hash.hash(key, self.size)

        if len(self.hash_table[index]) == 1 and self.hash_table[index][0][0] == key:
            return self.hash_table[index]
        else:
            for arr in range(len(self.hash_table[index])):
                if self.hash_table[index][arr][0] == key:
                    return self.hash_table[index][arr][1]
                    break


if __name__ == '__main__':
    HashTable1 = HashTable(4)  # create a hash table

    HashTable1.add("firstname", "shihara")  # add key value pairs for the hash table
    HashTable1.add("lastname", "dilshan")
    HashTable1.add("age", "umesha")
    HashTable1.add("address", "indeevari")
    HashTable1.add("Uni", "sliit")

    print(HashTable1.get("firstname"))  # get a value by key

    HashTable1.remove("Uni")

    print(HashTable1.get("Uni"))  # try to get a removed item by key

File: hash table/java/hashtable.js
/* hash table using javascript */

class Hash {
  //get a hash value for a given string
  static hash(input, maxSize) {
    let hash = 0;
    for (let i = 0; i < input.length; i++) {
      hash = hash + input.charCodeAt(i);
    }
    return hash % maxSize;
  }
}

class HashTable {
  constructor(sizeLimit) {
    this.hashTable = [];
    this.sizeLimit = sizeLimit;
  }

  //add a item to the hash table
  add(key, value) {
    let index = Hash.hash(key, this.sizeLimit);
    let inserted = false;

    if (this.hashTable[index] === undefined) {
      this.hashTable[index] = [[key, value]];
      inserted = true;
    } else {
      for (let i = 0; i < this.hashTable[index].length; i++) {
        if (this.hashTable[index][i][0] === key) {
          this.hashTable[index][i][1] = value;
          inserted = true;
        }
      }
    }

    if (inserted === false) {
      this.hashTable[index].push([key, value]);
    }
  }

  //remove a item to the hash table
  remove(key) {
    let index = Hash.hash(key, this.sizeLimit);

    if (this.hashTable[index] === undefined) {
        return undefined;
      }

    if (
      this.hashTable[index].length === 1 &&
      this.hashTable[index][0][0] === key
    ) {
      delete this.hashTable[index];
    } else {
      for (let i = 0; i < this.hashTable[index].length; i++) {
        if (this.hashTable[index][i][0] === key) {
          delete this.hashTable[index][i];
        }
      }
    }
  }

  //get a item from the hash table using the key
  get(key) {
    let index = Hash.hash(key, this.sizeLimit);

    if(this.hashTable[index] === undefined){
        return undefined;
    }

    if (
      this.hashTable[index].length === 1 &&
      this.hashTable[index][0][0] === key
    ) {
      return this.hashTable[index][0][1];
    } else {
      for (let i = 0; i < this.hashTable[index].length; i++) {
        if (this.hashTable[index][i][0] === key) {
          return this.hashTable[index][i][1];
        }
      }
    }
    return undefined;
  }
}

class Main {
  static main() {
    let HashTable1 = new HashTable(4);//create a hash table

    HashTable1.add("firstname", "shihara");//add key value pairs for the hash table
    HashTable1.add("lastname", "dilshan");
    HashTable1.add("age", "umesha");
    HashTable1.add("address", "indeevari");
    HashTable1.add("Uni", "sliit");

    console.log(HashTable1.get("firstname"));//get a value by key

    HashTable1.remove("Uni");

    console.log(HashTable1.get("Uni"));//try to get a removed item by key

  }
}

//do not edit after this line
Main.main();
